guide/miscellaneous/parsing-mention-arguments.md

```diff
@@ -89,7 +89,7 @@ const { MessageMentions: { USERS_PATTERN } } = require('discord.js');
 
 function getUserFromMention(mention) {
 	// The id is the first and only match found by the RegEx.
-	const matches = mention.match(USERS_PATTERN);
+	const matches = USERS_PATTERN.exec(mention);
```

Comment:
This is currently prone to the [statefulness of RegExp#exec](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec), meaning every second execution of the function will not yield any match. The patterns provided by [discord.js](https://github.com/discordjs/discord.js/blob/stable/src/structures/MessageMentions.js#L221) are saved on the object as instance, not provided as fresh instances through getters.

> JavaScript RegExp objects are stateful when they have the global or sticky flags set (e.g. /foo/g or /foo/y). They store a lastIndex from the previous match. Using this internally, exec() can be used to iterate over multiple matches in a string of text (with capture groups), as opposed to getting just the matching strings with String.prototype.match(). 

The following shows the issue:

```js
const USERS_PATTERN = /<@!?(\d{1,19})>/g;

function getUserFromMention(mention) {
  const matches = USERS_PATTERN.exec(mention);
  if (!matches) return;
  return matches[1];
}

console.log("---");
console.log("id: ", getUserFromMention("<@1>")); // 1
console.log("id: ", getUserFromMention("<@2>")); // undefined
console.log("id: ", getUserFromMention("<@3>")); // 3
console.log("id: ", getUserFromMention("<@4>")); // undefined
console.log("id: ", getUserFromMention("<@5>")); // 5
```

Reply:
I would suggest having a single RegExp instance if this is going to be used more than once... then simply reset that last index:

```js
import { MessageMentions } from "discord.js";

const USERS_REGEX = new RegExp(MessageMentions.USERS_PATTERN);

function getUserFromMention(mention) {
  USERS_REGEX.lastIndex = 0;
  const matches = USERS_PATTERN.exec(mention);
  if (!matches) return;
  return matches[1];
}

or create a new RegExp within that function, then no need to reset, (bad for GC)
```

Reply:
That's rather hacky, I'd prefer using <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/matchAll> over this